File: AtCoder/ABC128/D/d.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <map>
#include <vector>

#define rep(i, n) for (int i = 0; i < (int)(n); i++)
using namespace std;

int main() {

  int n, K;
  cin >> n >> K;
  vector<int> a(n);
  for (auto &e : a) {
    cin >> e;
  }
  int ans = 0;
  for (int k = 1; k <= K; k++) {
    for (int d = 1; d <= min(k, n); d++) { // d個取り出す
      auto e = k - d;                      // e個戻す
      if (d < e) continue;
      map<int, int> s;
      int sum = 0;
      int left = d - 1, right = n; // [0, left], [right, n - 1]
      rep(i, d) {
        s[a[i]] += 1;
        sum += a[i];
      }
      {
        int sub = 0, cnt = e;
        for (auto p : s) {
          sub += p.first * min(p.second, cnt);
          cnt -= p.second;
          if (cnt <= 0) break;
        }
        ans = max(ans, sum - sub);
      }
      while (left >= 0) {
        s[a[left]] -= 1;
        sum -= a[left];
        s[a[right - 1]] += 1;
        sum += a[right - 1];
        left--;
        right--;
        int sub = 0, cnt = e;
        for (auto p : s) {
          sub += p.first * min(p.second, cnt);
          cnt -= p.second;
          if (cnt <= 0) break;
        }
        ans = max(ans, sum - sub);
      }
    }
  }
  cout << ans << endl;

  return 0;
}

File: AtCoder/ABC128/D/d.nim
import strutils, math, sequtils, algorithm

let read = iterator: string {.closure.} =
  while true:
    for s in stdin.readLine.split:
      yield s

proc main() =
  let
    n, k = read().parseInt
    a = newSeqWith(n, read().parseInt)

  var ans = int.low
  for i in 0..<n:
    # [0, i)
    for j in i..n:
      # [j, n)
      # -10, 8, 2, 1, 2, 6
      if i + (n - j) <= k:
        let
          b = (a[0..<i] & a[j..^1]).sortedByIt(it)
          s = b.sum
        # echo i, " ", j, " ", b, " ", s
        for p in 0..(k - i - (n - j)):
          ans = max(ans, s - b[0..<min(b.len, p)].sum)
  echo ans
main()
